src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,259 @@ where
+            )?;
+            let pos_list = val.borrow().position_list();
+            let mut path = Vec::with_capacity(pos_list.len() - 1);
+            for &bit in &pos_list[..pos_list.len() - 1] {
+                // we do not include leaf bit here, so ignore last element returned by pos_list
+                path.push(Boolean::new_variable(
+                    ark_relations::ns!(cs, "path_bit"),
+                    || Ok(bit),
                     mode,
-                )?;
-                let r_hash = HGadget::OutputVar::new_variable(
-                    ark_relations::ns!(cs, "r_child"),
-                    || Ok(r),
+                )?)
+            }
+
+            let mut auth_path = Vec::with_capacity(val.borrow().auth_path.len());
+            for v in val.borrow().auth_path.iter() {
+                auth_path.push(TwoToOneHG::OutputVar::new_variable(
+                    ark_relations::ns!(cs, "auth_path_node"),
+                    || Ok(v.clone()),
                     mode,
-                )?;
-                path.push((l_hash, r_hash));
+                )?)
             }
-            Ok(PathVar { path })
+            Ok(PathVar {
+                path,
+                auth_path,
+                leaf_sibling,
+                leaf_position_bit,
+            })
         })
     }
 }
 
+impl<P, LeafH, TwoToOneH, ConstraintF> PathVar<P, LeafH, TwoToOneH, ConstraintF>
+where
+    P: Config,
+    LeafH: CRHGadget<P::LeafHash, ConstraintF>,
+    TwoToOneH: TwoToOneCRHGadget<P::TwoToOneHash, ConstraintF>,
+    ConstraintF: Field,
+{
+    /// Given a leaf, calculate the root as if the merkle has this leaf at this path.
+    fn calculate_root(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        let leaf_bytes = leaf.to_bytes()?;
+        let claimed_leaf_hash = LeafH::evaluate(leaf_hash_parameter, &leaf_bytes)?;
+        let leaf_sibling_hash = &self.leaf_sibling;
+
+        // calculate hash for the bottom non_leaf_layer
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is sibling if leaf_position_bit is 1, and leaf if bit is 0
+
+        let left_hash = self
+            .leaf_position_bit
+            .select(leaf_sibling_hash, &claimed_leaf_hash)?
+            .to_bytes()?;
+        let right_hash = self
+            .leaf_position_bit
+            .select(&claimed_leaf_hash, leaf_sibling_hash)?
+            .to_bytes()?;
+
+        let mut curr_hash =
+            TwoToOneH::evaluate_both(two_to_one_hash_parameter, &left_hash, &right_hash)?;
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+        let path_rev = {
```

Comment:
```suggestion
        let path_rev = self.path.iter().rev().to_vec()
```

Reply:
yes it works! (my compiler complains the `to_vec` so I will use `collect` instead. they should be equivalent lol)